File: bindings/go/dstu/memory_leak_test.go
package dstu

import (
	"bufio"
	"bytes"
	"io"
	"os"
	"runtime"
	"strconv"
	"strings"
	"testing"
)

// T-213: FFI memory-leak smoke test.
//
// This binding uses cgo to call into dstu-core-capi's C ABI, so native memory freed by
// dstu_*_free is allocated outside Go's own runtime heap - confirmed by reasoning, not
// guessed: cgo allocations made on the C side are invisible to runtime.ReadMemStats, which
// only accounts for memory the Go allocator itself manages. That's the same underlying
// category as the Java/.NET bindings' own T-213 tests in this batch (a native handle behind
// a Close()/finalizer, no Go-heap signal to observe it by) - see those two tests' doc
// comments for the two Windows-local measurement attempts (GC.GetTotalMemory-style JVM/CLR
// heap counters, then in-process working-set sampling) that were tried and rejected there
// before landing on the same mechanism this test uses: /proc/self/status's VmRSS, Linux-only,
// skipped elsewhere. Not re-attempting either rejected mechanism here is a deliberate
// consequence of this project's three-attempts rule (already spent on the same underlying
// question twice this batch), not an oversight.
//
// Not verified on this project's own Windows dev machine, matching the existing documented
// precedent for uacrypt_with_peak_rss's own Linux/macOS paths in the CLI test suite (reviewed,
// not run, before their first real CI confirmation) and this same T-213 batch's Java/.NET tests.
func currentVmRssBytes(tb testing.TB) int64 {
	tb.Helper()
	f, err := os.Open("/proc/self/status")
	if err != nil {
		tb.Fatalf("open /proc/self/status: %v", err)
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasPrefix(line, "VmRSS:") {
			fields := strings.Fields(line)
			kb, err := strconv.ParseInt(fields[1], 10, 64)
			if err != nil {
				tb.Fatalf("parse VmRSS value %q: %v", fields[1], err)
			}
			return kb * 1024
		}
	}
	tb.Fatal("VmRSS line not found in /proc/self/status")
	return 0
}

func runSecretstreamAndBoxLoop(tb testing.TB, key *SecretstreamKey, boxSecret *BoxSecretKey, boxPublic *BoxPublicKey, n int) {
	tb.Helper()
	for i := 0; i < n; i++ {
		var buf bytes.Buffer
		w, err := NewSecretStreamEncryptWriter(&buf, key, true)
		if err != nil {
			tb.Fatal(err)
		}
		if _, err := w.Write([]byte("leak-check chunk")); err != nil {
			tb.Fatal(err)
		}
		if err := w.Complete(); err != nil {
			tb.Fatal(err)
		}
		if err := w.Close(); err != nil {
			tb.Fatal(err)
		}

		r, err := NewSecretStreamDecryptReader(&buf, key, true)
		if err != nil {
			tb.Fatal(err)
		}
		got, err := io.ReadAll(r)
		if err != nil {
			tb.Fatal(err)
		}
		if string(got) != "leak-check chunk" {
			tb.Fatalf("secretstream round-trip mismatch: got %q", got)
		}
		if err := r.Close(); err != nil {
			tb.Fatal(err)
		}

		sealed, err := boxPublic.Seal([]byte("leak-check message"))
		if err != nil {
			tb.Fatal(err)
		}
		opened, err := boxSecret.Open(sealed)
		if err != nil {
			tb.Fatal(err)
		}
		if string(opened) != "leak-check message" {
			tb.Fatalf("box round-trip mismatch: got %q", opened)
		}
	}
}

func TestSecretstreamAndBoxLoopDoesNotLeak(t *testing.T) {
	if runtime.GOOS != "linux" {
		t.Skip("VmRSS-based leak check only runs on Linux - see file doc comment for why the Windows-local alternatives were rejected")
	}

	const warmup = 2000
	const n = 20000
	// Comfortable margin above normal churn but far below what N leaked handles would show at
	// this scale - same order of magnitude as this batch's Java/.NET thresholds.
	const maxAcceptableGrowthBytes = 8 * 1024 * 1024

	key, err := GenerateSecretstreamKey()
	if err != nil {
		t.Fatal(err)
	}
	defer key.Close()
	boxSecret, err := GenerateBoxSecretKey()
	if err != nil {
		t.Fatal(err)
	}
	defer boxSecret.Close()
	boxPublic := boxSecret.PublicKey()
	defer boxPublic.Close()

	runSecretstreamAndBoxLoop(t, key, boxSecret, boxPublic, warmup)
	runtime.GC()
	before := currentVmRssBytes(t)

	runSecretstreamAndBoxLoop(t, key, boxSecret, boxPublic, n)

	runtime.GC()
	after := currentVmRssBytes(t)
	growth := after - before
	if growth >= maxAcceptableGrowthBytes {
		t.Fatalf("VmRSS grew by %d bytes over %d iterations (threshold %d) - possible native handle leak", growth, n, maxAcceptableGrowthBytes)
	}
}

File: bindings/java/src/test/java/ua/dstucrypto/dstucore/MemoryLeakTest.java
package ua.dstucrypto.dstucore;

import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * T-213: FFI memory-leak smoke test. Unlike the other five direct-Rust bindings (Python/Node.js/
 * Ruby/PHP), this binding's {@link SecretStreamPushState}/{@link SecretStreamPullState} hold a raw
 * {@code long handle} (a boxed Rust pointer via JNI) with no {@code finalize()}/{@code Cleaner}
 * registered - freeing it is entirely {@link AutoCloseable#close}'s job. That makes a JVM-heap-based
 * measurement structurally blind to this leak class: confirmed empirically before writing this test
 * - {@code Runtime.totalMemory() - freeMemory()} showed no positive correlation at all between a
 * properly-closed loop and a deliberately-never-closed one (both noise, one even negative). A
 * process-RSS measurement via repeated {@code Get-Process -Id <pid>).WorkingSet64} sampling on this
 * project's own Windows dev machine was tried next and was too noisy to trust even with a warmup
 * pass and N=20000 (JIT/GC churn swamped the actual leak signal, which showed *smaller* growth for
 * the deliberately-leaked case than the properly-closed one in one run). Two failed local-measurement
 * attempts on Windows - per this project's three-attempts rule, this stops chasing a Windows-local
 * signal and uses the one mechanism that's actually low-noise: reading {@code /proc/self/status}'s
 * {@code VmRSS} line directly (no subprocess spawn, well-established stable kernel text convention -
 * same source this project's own {@code uacrypt_with_peak_rss} CLI test helper already trusts for
 * its Linux path). Linux-only; skipped elsewhere via {@link Assumptions#assumeTrue} rather than
 * asserting something this test can't reliably observe there - not verified on this project's own
 * Windows dev machine (matches the precedent {@code uacrypt_with_peak_rss}'s own doc comment already
 * sets: its Linux/macOS paths were reviewed, not run, before their first CI confirmation).
 */
class MemoryLeakTest {
    private static boolean isLinux() {
        return System.getProperty("os.name", "").toLowerCase(java.util.Locale.ROOT).contains("linux");
    }

    private static long currentVmRssBytes() throws IOException {
        List<String> lines = Files.readAllLines(Paths.get("/proc/self/status"));
        for (String line : lines) {
            if (line.startsWith("VmRSS:")) {
                String[] parts = line.trim().split("\\s+");
                // "VmRSS:", "<kB value>", "kB"
                return Long.parseLong(parts[1]) * 1024;
            }
        }
        throw new IOException("VmRSS line not found in /proc/self/status");
    }

    @Test
    void secretstreamAndBoxLoopDoesNotLeak() throws Exception {
        Assumptions.assumeTrue(isLinux(), "VmRSS-based leak check only runs on Linux - see class doc");

        int warmup = 2000;
        int n = 20000;
        // Comfortable margin above normal JVM/JIT/GC churn but far below what N leaked handles
        // (each holding at least a few dozen bytes of native Rust state) would show at this scale.
        long maxAcceptableGrowthBytes = 8L * 1024 * 1024;

        byte[] key = SecretStream.keygen();
        byte[] boxSecret = Box.keygen();
        byte[] boxPublic = Box.publicKey(boxSecret);

        runLoop(key, boxSecret, boxPublic, warmup); // trigger class-loading/JIT noise before measuring
        System.gc();
        Thread.sleep(200);
        long before = currentVmRssBytes();

        runLoop(key, boxSecret, boxPublic, n);

        System.gc();
        Thread.sleep(200);
        long after = currentVmRssBytes();
        long growth = after - before;
        assertTrue(
                growth < maxAcceptableGrowthBytes,
                "VmRSS grew by " + growth + " bytes over " + n + " iterations "
                        + "(threshold " + maxAcceptableGrowthBytes + ") - possible native handle leak");
    }

    private static void runLoop(byte[] key, byte[] boxSecret, byte[] boxPublic, int n) throws Exception {
        for (int i = 0; i < n; i++) {
            try (SecretStreamPushState push = new SecretStreamPushState(key)) {
                byte[] header = push.header();
                SecretStreamPushResult r = push.push(SecretStreamTag.MESSAGE, "leak-check chunk".getBytes());
                try (SecretStreamPullState pull = new SecretStreamPullState(key, header)) {
                    pull.pull(SecretStreamTag.MESSAGE.ordinal(), r.ciphertext(), r.authTag());
                }
            }

            byte[] sealed = Box.seal(boxPublic, "leak-check message".getBytes());
            byte[] opened = Box.open(boxSecret, sealed);
            assertArrayEquals("leak-check message".getBytes(), opened);
        }
    }
}
